### ISSUE TITLE
Improve the anti spam measures

### ISSUE THREAD

**Issue by dtdesign** [WoltLab/WCF#2692]
Evaluate options to combat spam even more, including but not limited to external services such as StopForumSpam. Whatever solution(s) we pick, they must be legal beyond doubt (GDPR/privacy shield) and should not require awkward contracts to be signed by each site owner.

ref https://community.woltlab.com/thread/264735-spam-problems/

**Comment by dtdesign** [comment]
I've laid out a few different plans on how to approach this, both considering performance bottlenecks and data privacy. I'm using "client" to refer to the "WoltLab Suite" installation. The performance was tested on an i5-9600K, 16 GB DDR4-2667 running PHP 7.3 on Windows 10/IIS 10, the files were loaded from a local NVMe drive (Samsung 970 Pro).

## Reasons against using the SFS API.
* Requires an individual contract between the website owner and SFS, because the transmitted data is subject to the GDPR.
* Hashing the email address using md5 is better than transmitting plain text, but from a technical standpoint it's far too easy to compute the original value.
* The username and ip address are always transmitted using its raw value and both can be either considered to be personal data or is actual personal data, e.g. the username is "<firstname> <lastname>".
* Requires the service to be available at the time of the request, including but not limited to system maintenance or issues with the network routing.

## Goal
Validate each registration against a reasonable amount of data:
* IPv4 and IPv6 (last 7 days)
* Email addresses (last 30 days)
* Usernames (last 30 days)
* Toxic domains (filtered Alexa Top 50k)

The IP, email and username data needs to be stored using hash sums, otherwise one would be semi-permanently storing personal data, which is a direct breach of the GDPR. This is especially true given that the data from SFS _can_ contain values that are accidentially submitted and that do belong to actual people (false-positive match).

## Scenario A: Download the raw data and read from it.
The most straightforward approach which uses the raw data to work with and the first that I tried. Although I assumed that this wasn't going to be fast, I wanted to actual try it to see how far I can go and if there are some simple tricks to reduce the effort it takes to process the data.

Results:
* The aforementioned data comes at a whopping 25megs of raw text.
* Just parsing the CSV from the email address list took over a second while the CPU sits at its peak turbo frequency.
* Running all validations, excluding the toxic domains (they're partial matches and thus would heavily skew the numbers), takes close to 3 seconds.

… and we just entered self-DoS territory.

## Scenario B: Use the non-CSV lists and read from them.
This is somewhat interesting, because it greatly reduces the file size, from previously ~11megs down to ~5megs, however you're losing the counters which give an idea on how often a particular value was encountered. This is a great metric when working with data sets spanning quite some period, because it provides a bit of context.

Results:
* The total data is down to ~12 megs and parsing these files on demand is still above 1 second.
* Extrapolating the performance on an average webhost yields something unacceptable.

More self-DoS. Yay.

## Scenario C: Read the raw data and throw it at MySQL.
There are two things that should be considered a fact when dealing with the average web host:
* PHP is capped at 30 secs.
* MySQL is mostly fast when reading, but don't dare to write a lot of data. I sometimes wonder if they use floppy disks as write cache?

I guess you can see where this is going, reading these large chunks of data has so many drawbacks that I won't even consider it past this point.

## Scenario D: Read the raw data from the past 24 hours only.
An intriguing approach: The client has to deal with just a small portion of data, reading it from the disk is still a bit too slow, but MySQL is able to handle this just fine.

Pro:
* Somewhat okay-ish time consumption when importing the data into MySQL.
* Loading the new data for each new day will incrementally build up a database of historic data.

Contra:
* New installations will start with a heavily limited set of data, loading larger chunks is simply not feasible on the average web host.
* Missing out on a day will leave holes in the data, allowing known spamers to bypass the filter. This is partly mitigated by the fact that those fake identities / email addresses are used for a few days, so there is a significant overlapping potential.

## Scenario E: Fetch and preprocess the data on woltlab.com
The data is fetched daily, cleaned up and converted into hashes and thus no hassle when temporarily fiddling with personal data in the client. Historic data can be made available for some time, allowing clients to catch up in case it missed some days.

Pro:
* Everything.

Contra:
* SFS uses the most restricted CC license that one can think of, including "No Derivative Works".
* **They do not bother replying to messages sent through their contact form, therefore we're unable to negotiate any kind of agreement.**

**Comment by Morik** [comment]
Why not using the 7 day zip archives?
They are about 1.5 MB each (ip, username and email) and missing one file should not have that much of an impact...

**Comment by dtdesign** [comment]
The 7 day archives are only of any use if you're trying to catch up. They do not qualify for daily updates due to both the redundancy and the increased amount of processing power required. I was not joking when I said that the average web host's MySQL performance is horrible, I wouldn't be surprised if some of them are really running on 5.4k rpm rusty disks.

**Comment by Morik** [comment]
I would keep track of every update and only push entry's of the 7 day dump that have a newer time stamp than the last update to the database.
This reduces the amount of Data significant and if the daily update failed once its simple to catch up.
The only downside is that you need to process about three times more data thru php (i just compared the 7 day ip dump to the 24h dump, other data sets have different metrics but ip has proven to be the best indicator) but if mysql is the bottleneck this shouldn't be to bad.

**Comment by dtdesign** [comment]
That's still around 8MB of uncompressed data, because we need to use the `_all` variants to obtain the timestamps ("last seen"). However, even if we filter those results, the amount of items is still very high.

I've used the `listed_email_7_all.txt` from yesterday and wrote a quick script to extract all items that have been added since `2019-02-19 00:00:01`. This has reduced the list from 54k items down to 15k, taking about 0.4 seconds, equalling about 1.5-2 seconds when running on the average web host.

If we're adding this up for all the data we need to work through, we're easily looking at about 5-7 seconds spend just fetching the data via HTTP (currently served via Cloudflare, thus I don't worry that much about latency right now) and performing the first round of processing in PHP.

That leaves us with about 20 seconds to throw an estimate of 40k rows at the database, that's more than 2k rows per second. I mean, there are quite a lot of databases around that even fail the SQL steps during the WCFSetup process due to the database being super slow.

Eventually, we would need to split this up into separate requests in order to finish the process without hitting the max_execution_time. That's why I'm so disappointed that their license is that restrictive, there would be so much that we can do to reduce the data, e. g. guranteeing that a specific record appears only once in a fixed interval.

Also, I just noticed that SoftCreatR is listed as a moderator on their forums, maybe he is able to get in touch with them, or at least ask them to reply to my message…

**Comment by Morik** [comment]
It seams that the reading of those larger chunks itself is not that much of a problem, so you could further decrease the number of rows if you run the cron up to 24 times a day and load by default the 24h dump with the filtering in place. Only if the last update (or newest entry in the database as you should include the number of reports as well as the last report for a better scoring result anyway) is older than 24h load the 7day dump.

Can't say much about Saschas current status at SFS, but i noticed him about this thread so he can give you more information's about this.

**Comment by SoftCreatR** [comment]
I'll contact Paul regarding your request.

As for this topic, here's a little helper method, that was used in our SFS plugin:

```
	/**
	 * Confidence score calculation
	 * Based on the blog post at http://amix.dk/blog/post/19588
	 *
	 * @author Sascha Greuel
	 */
	protected function confidence($reports = 0, $lastReport = 0) {
		$ups = $reports / 2;
		$gracePeriod = 7;
		
		$downs = floor((time() - $lastReport) / 86400) - $gracePeriod;
		
		if ($downs < 0) {
			$downs = 0;
		}
		
		$numReports = $ups + ($downs * 2);
		
		if (!$numReports) {
			return 0;
		}
		
		// Begin with the Wilson score interval
		// http://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval#Wilson_score_interval
		$z = 1; // 1.0 = 85%, 1.6 = 95%
		$phat = $ups / $numReports;
		$confidence = round(sqrt($phat + $z * $z / (2 * $numReports) - $z * (($phat * (1 - $phat) + $z * $z / (4 * $numReports)) / $numReports)) / (1 + $z * $z / $numReports) * 100, 2);
		
		return $confidence;
	}
```

**Comment by SoftCreatR** [comment]
Just spoke to Paul. You are free to use the databases as needed, regardless of the license.

So go forth and be awesome ;)

**Comment by stopforumspam** [comment]
correct, I've just spoke to @SoftCreatR . 
OT - apologies, i can't find your emails

**Comment by dtdesign** [comment]
That are some great news @SoftCreatR. However, we would need to receive such a permission from him directly for legal reasons. I'll submit our request via the contact form again and await the reply, that will do.

**Comment by SoftCreatR** [comment]
@dtdesign Did you see the comment from @stopforumspam ? That's him 😄 

> correct, I've just spoke to @SoftCreatR .
> OT - apologies, i can't find your email

**Comment by dtdesign** [comment]
I'm sorry, but that user account is the digital equivalent of a virgin, at least given what is publicly visible, anyone could own that account. I can see that this would be enough proof for an individual, but as a company you have to be more precise than that.

**Comment by SoftCreatR** [comment]
That's what i told him :D He will get in touch with you.

**Comment by stopforumspam** [comment]
Indeed this account does, 

https://twitter.com/StopForumSpam/status/1098704569037537280

**Comment by stopforumspam** [comment]
@dtdesign I'm really sorry but I can't find an email from you :(

**Comment by dtdesign** [comment]
@stopforumspam You can reach me via `ebert@woltlab.com` 🙂

**Comment by stopforumspam** [comment]
https://www.stopforumspam.com/forum/viewtopic.php?pid=49934#p49934

**Comment by dtdesign** [comment]
Very well, I guess we can count both twitter and the official forums, that's all I've asked for. I appreciate that you've took your time to help us in this case! ❤️

**Comment by dtdesign** [comment]
@SoftCreatR Looking at your code example for the confidence score, what exactly is `$n`?

**Comment by SoftCreatR** [comment]
It was supposed to be `$numReports`. Fixed the code above :)

**Comment by dtdesign** [comment]
https://github.com/WoltLab/blacklist contains the code that generates the delta updates.
